### DISCUSSION TITLE
Dynamic Regulation via Temperature Offset

### DISCUSSION THREAD

**Discussion by LennardPlay** [jmcollin78/versatile_thermostat#342] [Ideas]
Most TRVs / room theremostats publish both a CLIMATE:set_temperature and a CLIMATE:temperature_offset entity.
That opens up two possible methods for dynamic temperature regulation.

**Dynamically adjusting the CLIMATE:set_temperature within a few degrees around the VT.set_temperure, to reach the desired temperature more quickly and more reliably (like VT does right now)**
- Easy to understand
- Doesn't require a temperature_offset temperature
- Easy to troubleshoot in the HA history tab

**Keeping the CLIMATE:set_temperature the same as the VT:set_temperature and doing the dynamic regulation through the CLIMATE:temperature_offset (Telling the climate device, that the room is colder than it thinks)**
- Keeping the VT's and climate's set point the same (more elegant)
- No more confusion for family members or guests ("Why does the thermostat say 22°C when I just set it to 20°C?")
- Easily adjust the set temperature on the climate device itself without needing to go into manual mode. Temperature set points always are fully synchronized
- It needs to be tested how other climate devices react to frequent changes in CLIMATE:temperature_offset

Basically the 2 methods do the same thing. The only difference is where the addition of the regulation_offset + set_temperature takes place (in software vs in the climate device)

### An Example
**Mothod 1: (How VT works right now)**
VT:current_temp: 15°C
VT:set_temperature: 20°C
-> CLIMATE:set temperature: 21.5°C
-> Climate heats to 21.5°C internal temperature -> Room reaches the right temperature more quickly

**Mothod 2:**
VT:current_temp: 15°C
VT:set_temperature: 20°C
-> CLIMATE:set temperature: 20°C
-> CLIMATE:temperature_offset: -1.5°C
-> Climate heats to 21.5°C internal temperature -> Room reaches the right temperature more quickly

I think method 2 could bring an improvement to the VT integration and make it more useful. Let me know what you think of this idea :)

**Comment by maia** [comment]
I fully support this feature. My TRVs (Eve Home Thermo) unfortunately do not expose their offset value via the HA Homekit integration, but as there are many TRVs that expose this value, this feature would resolve lots of confusion.

**Comment by maia** [comment]
By the way, I would like to suggest to separate the short term self-regulation (which happens over minutes and hours depending on the setting) from the long term offset. This long term offset should be calculated over multiple days, persist across a HA restart and only update the offset value in the TRV every few weeks or so. Because the primary reason the offset needs to change is if the outdoor  temperature changes dramatically. If it's very, very cold outside, this increases the chance that the room isn't heated evenly, which leads to a higher offset, as the temperature measured at the TRV differs too much from the rest of the room.

Probably we should calculate this offset by a exponential moving average with a time frame of multiple days of the difference of temperatures measured by the room sensor and the sensor built into the TRV. This will prevent that changes in the target temperature effect this offset.

Also, this discussion might be of help: https://github.com/jmcollin78/versatile_thermostat/discussions/208

**Comment by jmcollin78** [comment]
All this discussion rely on the presence of the CLIMATE:temperature_offset method. I have never seen that before this post.

@LennardPlay, you said "Most TRVs / room theremostats publish [...] CLIMATE:temperature_offset entity". Can you explain why you have this feeling ?

I already explain this many times, but one of my goal is to have something the most simple as possible for maintainability reason. Have too many methods of regulation depending of the TRV model is a complexity I really don't want to deal with.

The fight for simplicity implies choices and sacrifices.

**Reply by LennardPlay** [reply]
It seems like my base assumption, that an CLIMATE:temperature_offset entity is pretty common, just isn't true. I was under the impression that this was a pretty standard feature. Every TRV or room thermostat that I've ever seen has some way to add/subtract an offset, for "calibration" purposes, but it seems like other climate integrations don't publish an input like that.
I completely understand your reasoning. My intension was to make this integration easier to understand instead of adding functionality for edgecases. 

Thank you for very much for considering my idea and writing this increadible integration

**Reply by maia** [reply]
@jmcollin78 I understand the wish for simplicity. If I had written an integration based on my use case, I would have only added a offset feature, as I consider this more relevant for TRVs as the TPI based self-regulation. It probably very much depends on your use case and experiences. My Eve Thermo TRVs have a configurable offset (unfortunately not exposed in the HA Homekit integration) and are pretty smart regarding the self-regulation. They do a great job deciding on how much to open the valve based on current and recent temperature and target temperature. Just the offset is a pain, as it's not a fixed value but very much dependant on the outdoor temperature (at least in my case).

Oh, and as I suggested the self-regulation a few months back, I gave you a link to https://better-thermostat.org/qanda/modes – this explains the use of the offset value. And this integration also lists the devices that support/expose the offset value here: https://better-thermostat.org/qanda/supported , see this excerpt:

> "However, if you would like to use the “local calibration mode”, Better Thermostat needs to support the thermostat integration of your device and your device needs to support this feature. Currently, these are the integrations compatible with local calibration mode:
> 
> Tado
> Zigbee2mqtt
> Deconz"

I don't know which TRVs are most frequently used, but zigbee2mqtt and deconz probably are pretty frequent setups on home assistant setups. And tado is probably a common player in german speaking countries, as they are they are based in Germany.

**Reply by LennardPlay** [reply]
It seems like we are describing pretty different things. My idea was to do the TPI regulation through the offset entity. 
I guess your use case can be setup pretty easily by a template automation: 
set CLIMATE: temperature_offset to (Room_temperature + Climate: temperature_offset - CLIMATE: current_tperature)

Now that I think of it, I might just try that myself and see how this performs with my TRVs in comparison to VTs regulation

**Comment by jmcollin78** [comment]
I'm totally lost of this subject... There is many discussion threads and if I try a synthetis :  one want a solution that is dedicated to some TRV, the other says it would have done something different, the third says that self-regulation as implemented is working fine, the 4th says that only offset from TRV is the best solution.

So, currently, I understand there is a point, but I have no clues to do something. What I would not do is clear : I would not implement something which is dependant of the device. The `over_climate` is there exactly to avoid that: the specific implementation is on the underlying TRV and VTherm is above this specific implementation.

If you think in certain case, BetterThermostat is better, implement the BetterThermostat and eventually (if you want to use the VTherm feature) add a VTherm around it with the `over_climate` type.

**Reply by maia** [reply]
Better Thermostat is a pain in many aspects, it completely messes with the valve control. So no, that's not an option. As @LennardPlay mentioned, the "easiest" way is to write an automation: use the HA statistics sensor to create a long term moving average of the room temp and also of the TRV measured temp, round the difference to 0,5°C and every day at 12:00 update the offset value in the TRV accordingly.

Besides that, I don't think there is a synthesis: the offset is a long term calibration, the self-regulation is a short term improvement. Some people might need both features, other people might just need one of the two.

**Reply by LennardPlay** [reply]
I agree. It seems like there are many different ideas, which might only apply to some cases and are not compatible with other setups (including my idea).
In my setup your regulation works quite well, as it is. My only grape with it, is that it goes into manual override mode, when I adjust the temperature on the TRV (which honestly doesn't happen often), but I think it's not really possible  to fix this.
I might just turn on the child lock function, so that all changes are made through VT

**Reply by jmcollin78** [reply]
Thank you @maia for this synthesis 😉

**Comment by bell07** [comment]
I think this discussion is the right place for my thoughts on the topic. I buyed ~10 different zigbee thermostats on second-hand marked and use them with VT in my home.

The most of the thermostats provide offset setting and I miss the support in VT for them. But the confusing thing is they are two different offset types in hardware. I try to explain with example: 
- Current external temperature: 22
- Target temperature: 21
- Current internal temperature: 20

I have often this situation my cast iron radiators overheating because internal thermostat temperature is low.

**1. internal sensor offset**
Device TS0601 (_TZE200_6rdj8dzm) provides this offset type as `climate.[name_of_the_thermostat]_offset` entity. If I set the offset to value 2, the Current internal temperature is adjusted to 22 and all is fine.

**2. heating power offset**
SPZB0001 (Eurotronic) provides this offset type as `climate.[name_of_the_thermostat]_offset` entity. In this case I need to set the value to -2. (or -1? I still experimenting). The displayed current internal temperature is unchanged at 20, but the thermostat does stop heating earlier.

Of course the temperature delta changes over time, therefore it would be nice VT is able to adjust them. Since I do not see any way to detect the offset type, I think this should be done by setting.

For 1: "internal sensor offset" we have already "Use internal temperature of the underlying" setting. It would be nice to have the option, to send the external - internal temperature difference to the offset entity instead of main setpoint.

For 2: "heating power offset" we have already "Self-regulation". It would be nice to have the option to send the target - regulated temperature difference to the offset entity instead of main setpoint.

Depending on hardware the user can decide how to use the offset entity. Maybe just a dropdown "Offset type" with values "None" (current way), "Internal temperature offset" (support for option 1) and "Calibration offset" (support for option 2)?

**Comment by jmcollin78** [comment]
Since the 7.xx release the calibration offset entity has been introduced with the regulation by direct control of the valve: https://github.com/jmcollin78/versatile_thermostat/blob/main/documentation/en/self-regulation.md#self-regulation-by-direct-valve-control

Is that not what you are expecting ?

**Reply by bell07** [reply]
The most of my thermostats does not support direct control by valve :-(.
 I looking for offset entity usage in control by climate

![Screenshot_20250330-121725_Trebuchet.png](https://github.com/user-attachments/assets/d86f7edd-15f5-4a8f-94a6-a3873757cb50)

Screenshot show my current setting for "type 2" adjustment
(Room external sensor is 21°C atm)

**Reply by Dome-Maurice** [reply]
Hi,

Not really :) At least I am not expecting that.

First, thanks for the great integration and all the work you’ve put into this.
Sorry for “grave-digging” this thread. I hope I don’t have a major misunderstanding about how VTherm and TRVs function. If so, please correct me or just ignore this ;) It’s not really my area of expertise.

I’ve been playing with VTherm over the last few days as a Better Thermostat replacement, and I’m really stuck trying to understand this:

It’s probably a misunderstanding on my side, but isn’t the internal offset in valve control (or over climate) completely useless since the control loop is handled by VTherm, and the TRV is basically just an actuator without its own calculations?
On a Side note:
If the valve opening is set to 0, the TRV’s own control loop takes over again. Is this expected behavior?
At least that’s how it works with the TRV I have, which results in a funny “rippling” while it cools down to the new setpoint, that is still set in the underlying by VTherm. Because the internal hysteresis is quite large (1°), it still heats in small bursts during the cooldown. ( Direct Valve over climate )

#### Without Direct Valve Control
From a theoretical standpoint, I understand the intention: adding the offset (between internal temperature and actual room temperature) directly to the setpoint. But in practice, there are situations where the setpoint limit of the underlying TRV prevents further heating. For example, when the TRV placement is bad and the internal temperature rises very quickly above the setpoint. In my case, the TRV reports 27°C internally, the setpoint limit is 30°C, and there is still a heating demand because the room temperature is 1–2°C below the desired setpoint of 23°.

I know I have a bad heat buildup around my TRVs, and 4 out of 5 don’t support a valve setpoint but all do support an offset. Wouldn’t it be more robust in these cases to dynamically adjust the internal offset instead of adding it to the underlying setpoint? It shouldn’t affect any of VTherm’s internal calculations, except for where the offset is applied: put into an entity or  adding to the underlying setpoint.( I am talking about the setting "Compensating for the Internal Temperature of the Underlying Equipment" not the "normal" calculated setpoint offset of VTherm )
With the added side benefit that the TRV would display a more realistic setpoint.

I think the slower, gradual offset calculation that others have proposed can ideally work, but not in every case, because the offset can heavily depend on room temperature and heating demand when there is heat buildup around the TRV. I Suppose i can run that Calculation every 5 min or so in an Automation but this would add an addional external change to the VTherm loop inputs).

Thanks in advance, and thanks again for the great integration!

**Comment by bell07** [comment]
Found this [automation](https://community.home-assistant.io/t/adjust-multiple-thermostats-offsets-with-external-temp-sensors/651141) and adjusted for my needs. My thermostats have STEP=1 for offset that is too rough. So I added different rounding based on above or bellow the target temperature.
My version: 
<details>
  <summary>My Offset Update</summary>

```
alias: Heizung Offset Update
description: ""
triggers:
  - trigger: time_pattern
    minutes: /15
conditions: []
actions:
  - repeat:
      for_each:
        - thermostat: climate.kuche_heizung
          calibration: number.kuche_heizung_lokaler_temperatur_offset
          sensor_exte: sensor.kuche_thermometer_temperatur_2
        - thermostat: climate.halle_heizung_bei_schreibtisch
          calibration: number.halle_heizung_bei_schreibtisch_lokaler_temperatur_offset
          sensor_exte: sensor.halle_thermometer_temperatur
        - thermostat: climate.halle_heizung_nordseite
          calibration: number.halle_heizung_nordseite_lokaler_temperatur_offset
          sensor_exte: sensor.halle_thermometer_temperatur
        - thermostat: climate.halle_heizung_vor_der_terrasse
          calibration: number.halle_heizung_vor_der_terrasse_lokaler_temperatur_offset
          sensor_exte: sensor.halle_thermometer_temperatur
        - thermostat: climate.tze200_kds0pmmv_ts0601_thermostat
          calibration: number.tze200_kds0pmmv_ts0601_zahl_temperature_calibration
          sensor_exte: sensor.buro_thermometer_temperatur
        - thermostat: climate.buro_alex_heizung_am_fenster
          calibration: number.buro_heizung_am_fenster_lokaler_temperatur_offset
          sensor_exte: sensor.buro_thermometer_temperatur
      sequence:
        - data:
            entity_id: "{{ repeat.item.calibration }}"
            value: >
              {% set s_calibration = repeat.item.calibration %}
              {% set s_sensor_extern = repeat.item.sensor_exte %}
              {% set s_thermostat = repeat.item.thermostat %}

              {# Get external temperature #}
              {% set extern = states(s_sensor_extern) | float %}

              {# Get current calibration settings #} 
              {% set cal_curr = states(s_calibration) | float %}
              {% set cal_min = state_attr(s_calibration, "min") | float %} 
              {% set cal_max = state_attr(s_calibration, "max") | float %}
              {% set cal_step = state_attr(s_calibration, "step") | float %}

              {# Get thermostat temperature and setpoint #} 
              {% set thermo_cur = state_attr(s_thermostat, "current_temperature") | float %} 
              {% set thermo_setpoint = state_attr(s_thermostat, "temperature") or 0 | float %}

              {# Current not rounded offset #}
              {% set cal_new = extern - thermo_cur %}

              {# Adjust value before rounding, based on above or bellow the setpoint #} 
              {% set expected_new = thermo_cur + cal_new %} 
              {% if expected_new >= thermo_setpoint %} 
              {%   set cal_new = cal_new + cal_step / 2 %} 
              {% else %}
              {%   set cal_new = cal_new - cal_step / 2 %} 
              {% endif %}

              {# New rounded setpoint #} 
              {% set cal_new = ( cal_new / cal_step ) | round * cal_step + cal_curr %} 

              {# min/max adjusted return value #} 
              {{ min(max(cal_new, cal_min), cal_max) }}
                           
          action: number.set_value
mode: single

```
  
</details>

This automation run in parallel to the VT and adjust the offsets. As you see, the logic is similar to "Use internal temperature of the underlying" logic (that is disabled in favor of automation), just the calculated offset is used in other way.
The automation covers the "1. internal sensor offset".
For 2. I have slightly modified second automation, that does work, but not accurately enough to be shown.

**Reply by jmcollin78** [reply]
This is interesting and could (should ?) be integrated directly in VTherm. What is a shame is that I already add it into the valve regulation by durect control. I certainly have to rework on this.

**Reply by jmcollin78** [reply]
I open a Feature request linked to this discussion.

**Reply by lwfitzgerald** [reply]
> I open a Feature request linked to this discussion.

I had a quick search through open issues, but I can't find any feature request for this to upvote?

Would it be possible to provide a link? I'd love to be able to use the `local_temperature_calibration` features on my Z2M-connected `Moes TRV801Z` TRVs:
<img width="1019" height="800" alt="image" src="https://github.com/user-attachments/assets/51495851-b22f-4814-a3af-4e5c6f22c458" />

Note that these TRVs do not have the opening/closing degrees entities that the Sonoff TRVZB:
<img width="1280" height="4704" alt="image" src="https://github.com/user-attachments/assets/61a5b635-dd6b-4696-a0c7-cc4cd1a073e7" />

**Comment by jkrenzer** [comment]
I am currently mostly using BT for my Zigbee TRVs and I appreciate the direct reflection of the settings on the display when the offset-path on control is used. This is certainly a plus for the acceptance factor and also the self-regulation is quite robust in case HA is unavailable. The TRVs just do their work. Over time the offset might become not accurate, but still regulation is largely working fine so no rants either. When using direct valve control, my TRVs only show a percentage, not a set temperature, which has been vetoed by the household as unacceptable so this option is out of the question for this setup here anyways.

Sadly BT lacks many features and is also quite intransparent (you only have like three choices for an algorithm, no temperature sensor monitoring/fallback, no setting maximum offset bounds, no settings for algorithms etc.) which makes it quite volatile in some non-standard room/heating/sensor layouts. Up to the point the thermostats switch off every some hours because the offsets accumulate to astronomical values as the algorithms run away. This is why I investigated further and just stumbled on VT.

TLDR: I think having the option to use a calibration entity in self-regulation as well has several benefits which might be worth considering.

**Comment by bell07** [comment]
This winter I played again with offset scripts and regulations to get the regulation stable in most rooms with my cheap devices collection from misc vendors. Got new insights and new ideas..

The automation I proposed above regulates offset as difference between internal and external thermometer. This is basically the same functionality like "Use internal temperature of the underlying", that works only on slow (less powerful) devices for me. 

Previously I distinguished between "1. internal sensor offset" and "2. heating power offset", in the meantime I think the "2." is just incomplete support for Eurotronic SPZB0001 devices. All other devices are "internal sensor offset" type. 

Now I think the best way is to implement the offset as "target setpoint extension", following the base formula:

`target setpoint = expected setpoint + offset.`

This could be combined with already implemented regulations.
For example:
Situation:
- External temp = 18
- Internal temp = 19
- VT setpoint (expected setpoint) = 21
- The existing slow regulation calculates the target setpoint 20.5 

Current behaviour:
- Set device setpoint to 20.5

The proposal:
- Calculate the offset: "expected (VT) setpoint - regulated (target) setpoint" 
  -  Offset = 21 - 20.5 = 0.5
  - Adjust value using "min offset", "max offset" and "offset step"
    -  In my example the offset remains 0.5
- Re-calculate target setpoint: "regulated (target) setpoint + offset"
  Target setpoint = 20.5 + 0.5 = 21
- Send the both values to the device: offset=0.5 and setpoint=21

This way the regulation is working and the device show the expected (21) target temperature as long as the offset can handle the regulation offset.

Please note the internal temperature depends on offset too.
Sending offset "0.5" to the climate changes the reported internal temperature from 19 to 19 + 0.5 => 19.5 

`internal temperature = reported internal temperature - offset.`
For this reason the regulation needs to be adjusted.
Internal temperature in regulation = reported internal temperature (19.5) - offset (0.5) = 19

One issue I need to mention is, the reported internal temperature is not updated immediately on some devices. 
The change 19.0 -> 19.5 can be changed in 0.1 steps over some time on some devices. Therefore the regulated value calculation should not appear directly after update.


PS: This idea got already a feature request #985, but is closed  because of no upvotes in summer months. @jmcollin78 any chance to reopen it? Maybe the new information above is helpful. If you need a "Beta-Tester", I'l do it.

**Reply by jkrenzer** [reply]
> `internal temperature = reported internal temperature - offset.`

Actually there are also TRVs which do not modify the reported local temperature but just adjust their internal regulation. See [this comment](https://github.com/Koenkk/zigbee2mqtt/issues/28703#issuecomment-3416412166) for an example.

**Comment by bell07** [comment]
I see. Your COZB0001 is "heating power offset" device. Reading trough https://github.com/Koenkk/zigbee2mqtt/issues/28649 I see confirmed, my SPZB0001 (one of the newer devices) seems to be wrongly recognized COZB0001. I do not use zigbee2mqtt.
@jkrenzer, in case the device is right recognized as COZB0001, does this change the offset entity behavior in HA?

**Reply by jkrenzer** [reply]
Yes, it is a COZB0001. It seems the manufacturer quite randomly mixed batches of hardware and firmware. This is why Z2M recognizes the different models by the firmware date. They both report the same hardware and model information.

As far as I have observed the behavior is inverted, as the reported temperature is not affected by the correction value, but the internal regulation of the TRV is. Using the same integration which regulates fine with some older SPZBs the COZB regulates *away* from the setpoint, thus staying well below or above it, depending on the initial room temperature.
